csrc/device_lower/pass/alias_memory.cpp

```diff
@@ -1385,62 +1409,370 @@ class AllocationReuseModifier : private kir::ExprMutator {
+//! area.
+//!
+//! [Syncs for Reused (Not Aliased) Shared Memory]
+//! We will need to ensure the block is synced to prevent a race hazard where
+//! C is written to before some threads have read B, leading to corruption of
+//! those threads. These syncs need to ensure that the last read that might
+//! overlap the new allocation has finished before the first write of the new
+//! allocation. In this case that could be accomplished with an arrive/wait
+//! barrier between time points c and d, but any synchronizations contained
+//! within that interval in the original kernel, including a __syncthreads(),
+//! should be preferred over inserting new syncs.
+//!
+//! [Aliased Allocations]
+//! We handle aliased allocations by using a single liveness interval spanning
+//! from the first write of the base Allocation to the last read of either that
+//! allocation or the last alias. In some cases, this could be suboptimal; for
+//! example if a large tensor is used briefly at the beginning and is aliased
+//! near the end of the Fusion, then with this approach we will not be able to
+//! re-use its memory in the middle, which might be wasteful.
+class StackBasedSharedMemAllocator : kir::IrVisitor {
  public:
+  StackBasedSharedMemAllocator(const AllocationInfoMap& allocation_info_map)
+      : allocation_info_map_(allocation_info_map) {}
+
   void allocate(std::vector<Expr*>& exprs) {
+    recordEvents();
+
+    // Traverse expressions: reclaim memory when we pass a blockSync, append to
+    // waiting_to_push_ when we pass an Allocate
     handle(exprs);
+
+    // This is done whenever we pass a syncing op, but we need to do it again in
+    // case there are some allocations waiting around to be allocated.
+    sortPushAndAssignWaiting();
   }
 
  private:
   using kir::IrVisitor::handle;
 
-  static Val* alignExpr(Val* addr, int64_t alignment = 16) {
-    if (alignment == 1) {
-      return addr;
+  void dispatch(Expr* expr) final {
+    auto pos_opt = allocation_info_map_.getScopeMap().getMaybeExprPos(expr);
+    if (!pos_opt.has_value()) {
+      // Aliased kir::Allocate exprs are replaced, so the replacements will not
+      // exist in the original ScopeMap. We handle those manually here.
+      position_++;
+    } else {
+      position_ = pos_opt.value();
     }
-    auto n_minus_one = IrBuilder::create<Val>(alignment - 1, DataType::Index);
-    return SimplifyingIrBuilder::bitwiseAndExpr(
-        SimplifyingIrBuilder::addExpr(addr, n_minus_one),
-        SimplifyingIrBuilder::bitwiseNotExpr(n_minus_one));
-  }
 
-  void handle(kir::Allocate* alloc) final {
-    if (alloc->memoryType() != MemoryType::Shared || alloc->alias()) {
-      return;
+    // Check whether this is a first write position for any allocations
+    auto it = first_write_positions_.find(position_);
+    if (it != first_write_positions_.end()) {
+      for (auto alloc_info : it->second) {
+        if (alloc_info->mem_type != MemoryType::Shared) {
```

Comment:
Can this condition be true given https://github.com/NVIDIA/Fuser/pull/703/files#diff-72b133c7b6642d8a36ca5e165ad9f3e10ee91b93ad4f0f4e67b3e83298a81e59R1683-R1685?

Reply:
You're right it should never be true. I'll remove it since we should have never inserted a non-shared allocation.